### ISSUE TITLE
[Ch 02][Notebook][Major] "정규방정식으로 한 번에 푼다"·(XᵀX)⁻¹ 서술 — 이 챕터는 p(10,000) > n(4,000) 이라 역행렬이 없음

### ISSUE THREAD

**Issue by modesta-jihye-lee** [yoon-gu/neuqes-101#92]
## 대상

- **챕터**: Ch 02 — sklearn Regression (첫 모델, 첫 Loss)
- **산출물**: Notebook (`02_sklearn_regression/02_sklearn_regression.ipynb` — `## 🚀 실습` 셀 `cell009`, `## ❓ FAQ` Q3 `cell020`)
- **기준 커밋**: `eed3dbe`
- **환경**: 정적 검토 + 실행본 `executed/02_sklearn_regression.ipynb` 대조 + 로컬 재현(sklearn 1.9.0, 동일 `seed=42`/`random_state=42`). **Colab 전체 실행 완료** (2026-08-15 · CPU 런타임 · 80초 · 에러 0건)

## 실제 결과

이 챕터는 학습 과정을 **정규방정식(normal equation)** 으로 설명하고, 역행렬이 들어간 식을 제시합니다.

`cell009`:
> 정답 $y$와의 MSE를 최소화하도록 $w, b$를 푸는 게 학습의 전부입니다 (**sklearn은 정규방정식으로 한 번에 풉니다**).

FAQ Q3 (`cell020`):
> `LinearRegression`은 SGD가 아니라 **정규방정식(normal equation)** 으로 한 번에 답을 내는 닫힌 형태 풀이입니다.
> $$w = (X^\top X)^{-1} X^\top y$$

### (a) 이 챕터의 셋업에서 그 역행렬은 존재하지 않습니다

같은 노트북 `cell008` 이 출력하는 shape 는 다음과 같습니다.

```
X_train: (4000, 10000), y_train: (4000,)
```

**feature 수 p = 10,000 > 샘플 수 n = 4,000** 입니다. $X^\top X$ 는 10,000 × 10,000 이지만 rank 는 최대 4,000 이므로 **특이행렬(singular)** 이고 $(X^\top X)^{-1}$ 는 존재하지 않습니다. 학습자가 이 식을 그대로 이 챕터 데이터에 적용하려 하면 막힙니다.

### (b) sklearn 이 실제로 쓰는 경로도 정규방정식이 아닙니다

`LinearRegression.fit` 은 역행렬을 만들지 않고 **최소제곱법(least squares)** 으로 풉니다.

| 입력 | 경로 |
|---|---|
| dense | `scipy.linalg.lstsq` (SVD 기반) |
| **sparse (이 챕터)** | `elif sp.issparse(X):` 분기 → `LinearOperator` + sparse 최소제곱(lsqr 계열, **반복 계산**) |

즉 "정규방정식으로 **한 번에**" 라는 서술은 계산 방법(닫힌 형태 공식 ↔ 반복 계산)과 수식(역행렬 존재 가정) 두 가지 모두에서 어긋납니다. p > n 에서는 오차를 0 으로 만드는 해가 무수히 많고, 이때 반환되는 것은 그중 **L2 norm 이 가장 작은 해**(minimum-norm solution)입니다.

### (c) 그 결과로 나온 `Train MSE: 0.0000` 이 노트북에서 설명되지 않습니다

`cell010` 출력:

```
Train MSE: 0.0000        ← 실제 5.06e-09, 학습 데이터를 사실상 완전히 보간
Test  MSE: 1.5565
Test  MAE: 0.9952
Test  R²:  0.2139
```

Train MSE 가 0 인 것은 모델이 좋아서가 아니라 **p > n 이라 학습 데이터를 정확히 통과하는 해가 존재**하기 때문입니다. 그런데 노트북은 이 출력에 아무 설명을 붙이지 않고, 이어지는 `## 🔬 해부` 셀은 **출력 범위** 이야기로만 넘어갑니다. 체크포인트 질문 4개에도 이 항목이 없습니다.

가이드 2쪽 Notebook 사례의 *"딱 떨어지는 수치(0.500, 정확히 0.5)는 의심하세요"* (같은 쪽 체크리스트 "출력 내용을 읽었는가")에 정확히 해당하는 자리입니다.

참고로 **WikiDocs 페이지에는 이 설명이 이미 있습니다** — `pages/02-sklearn_regression-practice.md` 의 "결과 해석":

> Train MSE가 0에 가까운 건 10,000차원 feature 가 4,000 샘플을 거의 완벽히 외운 과적합 신호입니다.

즉 페이지 검수자가 필요하다고 판단해 덧붙인 설명이 **정작 학습자가 실행하는 노트북에는 없습니다.**

## 기대 결과

세 곳을 한 사이클에서 정리합니다.

1. `cell009` — "sklearn 은 정규방정식으로 한 번에 풉니다" → **"sklearn 은 최소제곱법으로 한 번에 풉니다 — 딥러닝처럼 epoch 를 돌며 조금씩 배우지 않습니다"**
2. FAQ Q3 — 역행렬 식을 쓰려면 전제를 함께 적거나, 이 챕터 셋업에 맞는 서술로 교체. 제안:

   > `LinearRegression` 은 SGD 로 조금씩 학습하지 않고 **최소제곱법(least squares)** 으로 풉니다. 교과서의 정규방정식 $w = (X^\top X)^{-1} X^\top y$ 는 $X^\top X$ 가 가역일 때(대개 샘플 수 > feature 수)의 형태인데, 이 챕터는 feature 10,000 > 샘플 4,000 이라 $X^\top X$ 가 특이행렬이어서 역행렬이 없습니다. sklearn 은 역행렬 대신 최소제곱법(dense: SVD 기반 `lstsq`, sparse: lsqr 계열)으로 **L2 norm 이 가장 작은 해**를 찾습니다.

3. `cell010` 출력 아래(또는 `## 🔬 해부` 도입부)에 과적합 한 문단 추가 — 페이지의 "결과 해석" 문장을 노트북에도 반영하고, "Train MSE 0 은 성능이 아니라 p > n 의 결과" 를 명시. 체크포인트 질문에 한 항목 추가도 권장합니다.

## 재현 절차

```python
from sklearn.metrics import mean_squared_error
# 위 노트북 셀을 그대로 실행한 뒤
print(f"{mean_squared_error(y_train, y_pred_train):.3e}")   # 5.058e-09
print(X_train.shape)                                        # (4000, 10000) → p > n
```

```python
import inspect, sklearn.linear_model._base as base
print([l.strip() for l in inspect.getsource(base.LinearRegression.fit).split("\n")
       if "issparse" in l or "lstsq" in l])
```

## 로그

```
Train MSE 정확값: 5.058e-09   (노트북 표기 0.0000)
Train 최대 절대오차: 1.010e-03
X_train (4000, 10000)  →  n=4000 < p=10000, rank(X^T X) <= 4000 → 특이행렬

sklearn 1.9.0 LinearRegression.fit 분기:
    elif sp.issparse(X):            ← 이 챕터 (TF-IDF sparse)
    self.coef_, _, self.rank_, self.singular_ = linalg.lstsq(   ← dense 경로
```

## 전파 범위

| 산출물 | 상태 |
|---|---|
| `02_sklearn_regression/02_sklearn_regression.ipynb` `cell009`·`cell020` | ❌ 정규방정식·역행렬 서술 / 과적합 설명 없음 |
| `pages/02-sklearn_regression-practice.md` | △ 정규방정식 서술 있음 / 과적합 설명은 **있음** |
| `pages/02-sklearn_regression-variation.md`, `-wrapup.md` | ❌ 정규방정식 서술 |
| `book/chapters/ch02_sklearn_regression.tex` (5곳) | ❌ → EPUB 영향 |

## 심각도

**Major** — 계획서 §9 의 "잘못된 수식" 과 "중요한 설명 누락" 양쪽에 걸칩니다. 실행은 정상이지만, 학습자가 (1) 이 데이터에 존재하지 않는 역행렬 공식을 배우고 (2) `Train MSE 0.0000` 을 성능으로 오해할 수 있습니다.

## 재실행 여부

마크다운 셀 수정이므로 **노트북 재실행은 불필요**합니다. WikiDocs 재변환 → EPUB 재빌드가 필요합니다.

## 담당

- 수정 담당자: 미정
- 재검수자: 등록자 (@modesta-jihye-lee)
- 연결: 챕터 검수 이슈 #26

**Comment by yoon-gu** [comment]
커밋 \`65450eb\` 으로 잔여분을 정리하고 닫습니다.

PR #289 가 FAQ Q3 와 실습 셀을 정정했지만, 같은 장 안의 다른 서술 5종이 옛 용어를 그대로 쓰고 있었습니다. Q3 는 "p(10,000) > n(4,000) 이라 역행렬이 존재하지 않는다"고 말하는데 Q2 는 "정규방정식(closed-form)으로 푼다"고 말하는 자기모순 상태였습니다.

정리한 곳 — 7개 파일 12곳:

| 위치 | 이전 → 이후 |
|---|---|
| 노트북 cell19 FAQ Q2 | `정규방정식(closed-form)으로 풀기 때문에` → `최소제곱법으로 한 번에 풀기 때문에` |
| 노트북 cell20 삽질 코너 힌트 | `닫힌 형태 풀이에서` → `최소제곱법 해에서` |
| `pages/02-sklearn_regression-variation.md` :29 :41 | `닫힌 형태 풀이의 결과를` → `최소제곱법 해를`, `정규방정식으로 풀기에` → `최소제곱법으로 풀기에` |
| `02_sklearn_regression/README.md` :9 | `닫힌 형태 풀이 (정규방정식)` → `최소제곱법 풀이 (p > n 이라 정규방정식의 역행렬은 없음)` |

`pages/` 의 2곳은 노트북에 대응 셀이 없는 **페이지 전용 문장**이라 재변환으로는 따라오지 않아 직접 고쳤습니다.

반영 경로는 노트북 소스 → `executed/` → `pages/` → `book/chapters/ch02_sklearn_regression.tex` → `_drafts/_build_ch02.py` 전 계층입니다. 마크다운 셀만 손대 **재실행은 불필요**했고, 생성기까지 고쳐 재생성 시 되돌아가지 않게 했습니다. 노트북 lint 92개 clean.

Q3 본문의 `교과서의 정규방정식 $w = (X^\top X)^{-1} X^\top y$` 서술과 tex 색인 항목은 정정된 설명의 일부라 그대로 두었습니다.
